### PR TITLE
sstable: add more commentary on synthetic suffix replacement

### PR DESCRIPTION
**Pull request by msbutler**
See individual commits

### CODE REVIEW

File: sstable/block.go [hunk 1]
```diff
@@ -1008,7 +1008,27 @@ func (i *blockIter) SeekLT(key []byte, flags base.SeekLTFlags) (*InternalKey, ba
 		i.offset = decodeRestart(i.data[i.restarts+4*(index-1):])
 		if index < i.numRestarts {
 			targetOffset = decodeRestart(i.data[i.restarts+4*(index):])
+
```
Comment (msbutler) [inline sstable/block.go:1011]: If you like this approach better, I'll add a big comment explaining what's going on here.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3300)* status: 0 of 7 files reviewed, 2 unresolved discussions (waiting on @msbutler and @sumeerbhola)

___
*[`sstable/suffix_rewriter.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqdQHppC_NR3PvwDYk9:-NqdQHppC_NR3PvwDYkA:bvdkyvc) ([raw file](https://github.com/cockroachdb/pebble/blob/6b0359662b26ddb4c8583832081f5d809ad5bb01/sstable/suffix_rewriter.go#L204)):*
> ```Go
> 			return err
> 		}
> 		if err := iter.init(r.Compare, r.Split, inputBlock, r.Properties.GlobalSeqNum, false /* hideObsoletePoints */, nil /* syntheticSuffix */); err != nil {
> ```

Sorry to be a pain but could you leave these call sites as they are? I am moving them to an `IterTransforms` struct which contains `HideObsoletePoints` and `SyntheticSuffix` (and the "global" sequence number), and this would just create merge conflicts. It will be addressed very soon.

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3300)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @10, @20, @msbutler, and @sumeerbhola)

___
*[`sstable/block.go` line 305 at r1](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nqh1nPAD4U5M-E3Am1X:-NqhLib_5A-siRwQGXWL:b-m349dj) ([raw file](https://github.com/cockroachdb/pebble/blob/52e000d13771e34ad535e843a7e644b5b68d5b10/sstable/block.go#L305)):*
> Are we only relying on the key stability guarantee when processing rangedels and rangekeys?

Yeah, I believe that's true

> What about rangekeys? If someone wrote a RANGEKEYSET `[b, g)@10`, and now we want to replace suffix `@10` with suffix `@20`, how is that accomplished?

Also, what happens when a range key fragment has a suffix in either the start or end key? That can happen in ordinary non-ingested sstables due to fragmentation, but maybe it cannot happen when we produce these sstables?

Or maybe are there no range keys in these sstables at all because they're produced without version history?

**Comment by RaduBerinde** [comment]
*[`sstable/block.go` line 305 at r1](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nqh1nPAD4U5M-E3Am1X:-NqhVoVlC31cjJqTpoji:b-6jcyed) ([raw file](https://github.com/cockroachdb/pebble/blob/52e000d13771e34ad535e843a7e644b5b68d5b10/sstable/block.go#L305)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

> Are we only relying on the key stability guarantee when processing rangedels and rangekeys?

Yeah, I believe that's true

> What about rangekeys? If someone wrote a RANGEKEYSET `[b, g)@10`, and now we want to replace suffix `@10` with suffix `@20`, how is that accomplished?

Also, what happens when a range key fragment has a suffix in either the start or end key? That can happen in ordinary non-ingested sstables due to fragmentation, but maybe it cannot happen when we produce these sstables?

Or maybe are there no range keys in these sstables at all because they're produced without version history?
</blockquote></details>

Yeah we don't allow rangekeys in external ingest for now (we won't need to for the first OR version).

**Comment by RaduBerinde** [comment]
*[`sstable/suffix_rewriter.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqdQHppC_NR3PvwDYk9:-Nqmd3DQ3aRQCNXChmMt:b-lxt62f) ([raw file](https://github.com/cockroachdb/pebble/blob/6b0359662b26ddb4c8583832081f5d809ad5bb01/sstable/suffix_rewriter.go#L204)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sorry to be a pain but could you leave these call sites as they are? I am moving them to an `IterTransforms` struct which contains `HideObsoletePoints` and `SyntheticSuffix` (and the "global" sequence number), and this would just create merge conflicts. It will be addressed very soon.
</blockquote></details>

https://github.com/cockroachdb/pebble/pull/3310

**Comment by RaduBerinde** [comment]
*[`sstable/suffix_rewriter.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqdQHppC_NR3PvwDYk9:-NqmfhQF5yaJVry5Ali1:b4qxvct) ([raw file](https://github.com/cockroachdb/pebble/blob/6b0359662b26ddb4c8583832081f5d809ad5bb01/sstable/suffix_rewriter.go#L204)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

https://github.com/cockroachdb/pebble/pull/3310
</blockquote></details>

The PR above also adds a check that we never try to iterate through range dels or range keys when we have a synthetic suffix.

**Review by msbutler — COMMENTED** [review]
Removed last commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3300#-:-NqnVbT53INgwG886Pks:b47h2nm)* status: 0 of 7 files reviewed, 5 unresolved discussions (waiting on @10, @20, @msbutler, @RaduBerinde, and @sumeerbhola)

___
*[`sstable/block.go` line 305 at r1](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nqh1nPAD4U5M-E3Am1X:-NqnYD-2Euo33mT67exy:b-164t4w) ([raw file](https://github.com/cockroachdb/pebble/blob/52e000d13771e34ad535e843a7e644b5b68d5b10/sstable/block.go#L305)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yeah we don't allow rangekeys in external ingest for now (we won't need to for the first OR version).
</blockquote></details>

Added more commentary on this. 
___
*[`sstable/block.go` line 1011 at r2](https://reviewable.io/reviews/cockroachdb/pebble/3300#gh-1489758567:-NqnYK1zDWAIZHYyQmJb:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/99105b73ef641db7988111667310fd40dc806a0c/sstable/block.go#L1011)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Yes, I have a preference for this since it sets up everything correctly for later iteration.
</blockquote></details>

Done.
___
*[`sstable/block.go` line 1016 at r2](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqhABC51TwMS1AghUBo:-NqnYL3W8-8mTm2L_4UW:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/99105b73ef641db7988111667310fd40dc806a0c/sstable/block.go#L1016)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we don't need to set `i.nextOffset`. `i.readEntry()` will set that.
</blockquote></details>

Done.
___
*[`sstable/block.go` line 1025 at r2](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqhAGCH3c_qBGDH5lSY:-NqnYLtb-41wY3IKSR8q:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/99105b73ef641db7988111667310fd40dc806a0c/sstable/block.go#L1025)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

index+1 may be out of bounds.
</blockquote></details>

Done.
___
*[`sstable/block.go` line 1030 at r2](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqhASzr0JJzt6nOxycG:-NqnVSX4EHa8tIbX-Dpr:bmobqgs) ([raw file](https://github.com/cockroachdb/pebble/blob/99105b73ef641db7988111667310fd40dc806a0c/sstable/block.go#L1030)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think these blocks can be rearranged to be clearer (some of this predates your changes, but we need to simplify to compensate for the added complexity)

Let's lift the index==0 case before, so we can delay definition of the `targetOffset` and can clarify what it really means. And then we can unindent the main case. Something like:

```
if index == 0 {  
  if i.syntheticSuffix != nil {  
   // The binary search was conducted on keys without suffix replacement,  
   // implying the first key in the block may be less than the search key. To   // double check, get the first key in the block with suffix replacement   // and compare to the search key. Consider the following example: suppose   // the user searches with a@3, the first key in the block is a@2 and the   // block contains a suffix replacement rule of 4. Since a@3 sorts before   // a@2, the binary search would return index==0. Without conducting the   // suffix replacement, the SeekLT would incorrectly return nil. With   // suffix replacement though, a@4 should be returned as a@4 sorts before   // a@3.   ikey, lazyVal := i.First()  
   if i.cmp(ikey.UserKey, key) < 0 {  
    return ikey, lazyVal  
   }  
  }  
  // If index == 0 then all keys in this block are larger than the key  
  // sought.  i.offset = -1  
  i.nextOffset = 0  
  return nil, base.LazyValue{}  
}  
  
// INVARIANT: index > 0  
  
// Ignoring suffix replacement, we will search between the restart at index-1  
// and the restart at index, for the first key >= key, and then on finding it,  
// return i.Prev(). We need to know when we have hit the offset for index, since  
// then we can stop searching. targetOffset encodes that offset for index.  
targetOffset := i.restarts  
  
i.offset = decodeRestart(i.data\[i.restarts+4\*(index-1):\])  
if index < i.numRestarts {  
  targetOffset = decodeRestart(i.data\[i.restarts+4\*(index):\])  
  if i.syntheticSuffix != nil {  
   // It is possible that we should be searching between the restart at index  
   // and that at index+1. TODO: add longer comment with example  
   naiveOffset := i.offset  
   // Shift up to the original binary search result and decode the key.  
   i.offset = targetOffset  
   i.readEntry()  
   i.decodeInternalKey(i.key)  
   i.maybeReplaceSuffix(false /\* allowInPlace \*/)  
  
   // If the binary search point is actually less than the search key, post  
   // replacement, bump the target offset.   if i.cmp(i.ikey.UserKey, key) < 0 {  
    i.offset = targetOffset  
    if index+1 < i.numRestarts {  
     targetOffset = decodeRestart(i.data\[i.restarts+4\*(index+1):\])  
    } else {  
     targetOffset = i.restarts  
    }  
   } else {  
    i.offset = naiveOffset  
   }  
  }  
}
```
</blockquote></details>

Done, with added commentary for suffix replacement.
___
*[`sstable/suffix_rewriter.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/pebble/3300#-NqdQHppC_NR3PvwDYk9:-NqnVXVm65mcLNnIvEZ2:bc4am8q) ([raw file](https://github.com/cockroachdb/pebble/blob/6b0359662b26ddb4c8583832081f5d809ad5bb01/sstable/suffix_rewriter.go#L204)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The PR above also adds a check that we never try to iterate through range dels or range keys when we have a synthetic suffix.
</blockquote></details>

Removed the third commit from this pr. Will deal with later.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3300#-:-NqooYSs09MXb31mdqY2:bg7igkv)* status: 1 of 7 files reviewed, 5 unresolved discussions (waiting on @10, @20, @msbutler, and @RaduBerinde)

**Comment by RaduBerinde** [comment]
*[`sstable/block.go` line 438 at r4](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nr1inlP2rX3ULhUI3sm:-Nr1inlQ-aLM0ye3CpOg:b2jnkxr) ([raw file](https://github.com/cockroachdb/pebble/blob/c81fbe1f03d31a396f4376a506c6b09b1a9b0622/sstable/block.go#L438)):*
> ```Go
> 	//
> 	// (1) no two keys in the block share the same prefix.
> 	// (2) pebble.Compare(keyPrefix{replacementSuffix},keyPrefix{originalSuffix}) < 0
> ```

we also allow `originalSuffix` to be empty, correct?

**Comment by msbutler** [comment]
*[`sstable/block.go` line 438 at r4](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nr1inlP2rX3ULhUI3sm:-Nr1pMEO4JcwcYkqmaiV:bt6xfj9) ([raw file](https://github.com/cockroachdb/pebble/blob/c81fbe1f03d31a396f4376a506c6b09b1a9b0622/sstable/block.go#L438)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

we also allow `originalSuffix` to be empty, correct?
</blockquote></details>

Ah, i should have clarified given our conversation.  Yes, that is true.

And to clarify, keys with an empty suffix sort before keys with a suffix
`a@,a@5,a@4`

In other words, conducting suffix replacement on a key without a suffix would make it less than the original value, which contradicts this invariant.

This makes me wonder if all my added off-by-one logic breaks down when we conduct suffix replacement on keys without a suffix.

**Comment by msbutler** [comment]
*[`sstable/block.go` line 438 at r4](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nr1inlP2rX3ULhUI3sm:-Nr1u5vK-zFRylniL-T_:b-g75072) ([raw file](https://github.com/cockroachdb/pebble/blob/c81fbe1f03d31a396f4376a506c6b09b1a9b0622/sstable/block.go#L438)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Ah, i should have clarified given our conversation.  Yes, that is true.

And to clarify, keys with an empty suffix sort before keys with a suffix
`a@,a@5,a@4`

In other words, conducting suffix replacement on a key without a suffix would make it less than the original value, which contradicts this invariant.

This makes me wonder if all my added off-by-one logic breaks down when we conduct suffix replacement on keys without a suffix.
</blockquote></details>

note: the only keys without a suffix are index keys

**Comment by RaduBerinde** [comment]
*[`sstable/block.go` line 438 at r4](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nr1inlP2rX3ULhUI3sm:-Nr1xkNn9o8ky6I55K0K:b-17oyxu) ([raw file](https://github.com/cockroachdb/pebble/blob/c81fbe1f03d31a396f4376a506c6b09b1a9b0622/sstable/block.go#L438)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

note: the only keys without a suffix are index keys
</blockquote></details>

In the meta test we do have keys without prefix and it seems to work fine.

**Comment by msbutler** [comment]
*[`sstable/block.go` line 438 at r4](https://reviewable.io/reviews/cockroachdb/pebble/3300#-Nr1inlP2rX3ULhUI3sm:-Nr5zLv1DJ-An7_Xmd6u:b-5tjvtk) ([raw file](https://github.com/cockroachdb/pebble/blob/c81fbe1f03d31a396f4376a506c6b09b1a9b0622/sstable/block.go#L438)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In the meta test we do have keys without prefix and it seems to work fine.
</blockquote></details>

Alrighty, i feel a bit better about this after adding a few unit test cases here
https://github.com/cockroachdb/pebble/pull/3320

I think this all works is because we only need to trigger off by one handling when (integer cmp)  origSuffix < > searchkeySuffix <> replacementSuffix (i.e. in the searchkeySuffix is in between the other two suffixes). This condition will never trigger because the originalSuffix is the max suffix, the empty suffix.
